driver_name_cap/src/disco.lua

```diff
@@ -0,0 +1,88 @@
+local json = require 'dkjson'
+local log = require 'log'
+
+local function generate_block(size)
+    local hex_digits = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'}
+    local ret = ''
+    for _ = 1,size,1 do
+        ret = ret .. hex_digits[math.random(1, 16)]
+    end
+    if size < 12 then
+        ret = ret .. '-'
+    end
+    return ret
+end
+
+local function new_uuid()
+    return generate_block(8) ..
+            generate_block(4) ..
+            generate_block(4) ..
+            generate_block(4) ..
+            generate_block(12)
+end
+
+--- Add a new device to this driver
+---
+---@param driver Driver The driver instance to use
+---@param device_number number|nil If populated this will be used to generate the device name/label if not, `get_device_list`
+---                                     will be called to provide this value
+local function add_device(driver, device_number)
+    log.trace('add_device')
+    if device_number == nil then
+        log.debug('determining current device count')
+        local device_list = driver.device_api.get_device_list()
+        device_number = #device_list
+    end
+    local device_name = 'Device ' .. device_number
+    log.debug('adding device ' .. device_name)
+    local device_id = new_uuid()
+    local device_info = {
+        type = 'LAN',
+        deviceNetworkId = device_id,
+        label = device_name,
+        profileReference = 'driver_name.v1',
+        vendorProvidedName = device_name,
+    }
+    local device_info_json = json.encode(device_info)
+    local success, msg = driver.device_api.create_device(device_info_json)
+    if success then
+        log.debug('successfully created device')
+        return device_name, device_id
+    end
+    log.error(string.format('unsuccessful create_device %s', msg))
+    return nil, nil, msg
+end
+
+--- A discovery pass that will discover exactly 1 device
+--- for a driver. I any devices are already associated with
+--- this driver, no devices will be discovered
+---
+---@param driver Driver the driver name to use when discovering a device
+---@param opts table the discovery options
+---@param cont function function to check if discovery should continue
+local function disco_handler(driver, opts, cont)
```

Comment:
@FreeMasen For my own understanding, I wrote a process of events:
- Profile switches, aka a call to try_update_metadata occurs
- The hub has to re-discover the device with the driver that has the updated profile embedded in it?
- For that discovery process, we have implemented our own handler for discovery
- At this point, we have a specific customCapability that includes the saved information from before? Which we then emit in the device_added lifecycle event? 

- Therefore, we can "save" a new profile name in this custom capability, and emit it in device_added so that the preference can be updated at this point. Does that sound correct?

There are quite a few question marks here; would you mind helping clear up some of this? Hopefully my explanation has given a clear-ish idea of where my head is at right now on this topic.